nodejs/eks/cni.ts

```diff
@@ -0,0 +1,160 @@
+    if (args.warmEniTarget !== undefined) {
+        env.push({name: "WARM_ENI_TARGET", value: args.warmEniTarget.toString()});
+    }
+    if (args.warmIpTarget !== undefined) {
+        env.push({name: "WARM_IP_TARGET", value: args.warmIpTarget.toString()});
+    }
+    // Return the computed YAML.
+    return cniYaml.map(o => `---\n${jsyaml.safeDump(o)}`).join("");
+}
+
+function applyCniYaml(yamlPath: string, args: CniInputs) {
+    // Dump the kubeconfig to a file.
+    const tmpKubeconfig = tmp.fileSync();
+    fs.writeFileSync(tmpKubeconfig.fd, args.kubeconfig);
+
+    // Compute the required CNI YAML and dump it to a file.
+    const tmpYaml = tmp.fileSync();
+    fs.writeFileSync(tmpYaml.fd, computeCniYaml(yamlPath, args));
+
+    // Call kubectl to apply the YAML.
+    childProcess.execSync(`kubectl apply -f ${tmpYaml.name}`, {
+        env: { ...process.env, "KUBECONFIG": tmpKubeconfig.name },
+    });
+}
+
+/**
+ * Cni manages the configuration of the Amazon VPC CNI plugin for Kubernetes by applying its YAML chart. Once Pulumi is
+ * able to programatically manage existing infrastructure, we can replace this with a real k8s resource.
+ */
+export class Cni extends pulumi.dynamic.Resource {
+    constructor(name: string, kubeconfig: pulumi.Input<any>, args?: CniOptions, opts?: pulumi.CustomResourceOptions) {
+        // Check to ensure that kubectl is installed, as we'll need it in order to deploy k8s resources below.
+        try {
+            which.sync("kubectl");
+        } catch (err) {
+            throw new pulumi.RunError("Could not find kubectl. See https://kubernetes.io/docs/tasks/tools/install-kubectl/#install-kubectl for installation instructions.");
+        }
+
+        const yamlPath = path.join(__dirname, "cni", "aws-k8s-cni.yaml");
+
+        const provider = {
+            check: (state: any, inputs: any) => Promise.resolve({inputs: inputs, failedChecks: []}),
+            diff: (id: pulumi.ID, state: any, inputs: any) => Promise.resolve({}),
+            create: (inputs: any) => {
+                applyCniYaml(yamlPath, <CniInputs>inputs);
+                return Promise.resolve({id: crypto.randomBytes(8).toString("hex"), outs: {}});
+            },
+            update: (id: pulumi.ID, state: any, inputs: any) => {
+                applyCniYaml(yamlPath, <CniInputs>inputs);
+                return Promise.resolve({outs: {}});
+            },
+            read: (id: pulumi.ID, state: any) => Promise.resolve({id: id, props: state}),
+            delete: (id: pulumi.ID, state: any) => Promise.resolve(),
+        };
+
+        args = args || {};
+        super(provider, name, {
+            kubeconfig: pulumi.output(kubeconfig).apply(JSON.stringify),
+            nodePortSupport: args.nodePortSupport,
+            customNetworkConfig: args.customNetworkConfig,
+            externalSnat: args.externalSnat,
+            warmEniTarget: args.warmEniTarget,
+            warmIpTarget: args.warmIpTarget,
+        }, opts);
```

Comment:
Looks like you could also do something like `{ kubeconfig: pulumi.output(kubeconfig).apply(JSON.stringify, ...args}`?

Reply:
Yes: we do not want any extra kvps that may be present in `args` to be passed down to the engine.